### PR TITLE
[MM-24436]- Add a threshold from bottom for new messages toast

### PR DESCRIPTION
**Pull request by sridhar02**
#### Summary
Changed the buffer value to be considered from bottom for the toast message

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/14904
JIRA https://mattermost.atlassian.net/browse/MM-24436

### CODE REVIEW

**Comment by sudheerDev** [comment]
@sridhar02 Hey, BUFFER_TO_BE_CONSIDERED_BOTTOM is used for determining if the scroll view is at the bottom. we give a 10px buffer and still consider it as bottom. 

This value is used for various states like if user is at the bottom and he receives a new message, view is auto scrolled to the bottom. Now with this change even if user is 999px he is autoscrolled to the bottom. 

The way to implement is to have a look at `toast_wrapper` code which has a mock for archive toast which takes 1000px value into consideration. 

Hope this helps. You can reach out to me incase of questions

**Comment by sridhar02** [comment]
@sudheerDev can u recheck the PR, i have changed the logic to ``toast_wrapper`` component and changed related test cases related to toast_wrapper.

File: components/toast_wrapper/toast_wrapper.jsx [hunk 1]
```diff
@@ -84,7 +84,7 @@ class ToastWrapper extends React.PureComponent {
         }
 
         // show unread toast on mount when channel is not at bottom and unread count greater than 0
-        if (typeof showUnreadToast === 'undefined' && props.atBottom !== null) {
+        if (typeof showUnreadToast === 'undefined' && props.atBottom !== null && props.initScrollOffsetFromBottom > 1000) {
```
Comment (sudheerDev) [inline components/toast_wrapper/toast_wrapper.jsx:87]: We want toast value to be either `true` or `false`, but for assigning the value to `showUnreadToast` we need to wait for the scroll position to be set i.e wait for view to scroll to position for new message etc hence we set the first value at this condition.

`if (typeof showUnreadToast === 'undefined' && props.atBottom !== null) {`

This is true when a component is mounted and scroll position is set for the first time because `props.atBottom` will be not be null anymore. But if change it to 

`if (typeof showUnreadToast === 'undefined' && props.atBottom !== null && props.initScrollOffsetFromBottom > 1000) {`
Then it will not set the initial value of `showUnreadToast` anymore.

Instead this should be
```
if (typeof showUnreadToast === 'undefined' && props.atBottom !== null) {
    showUnreadToast = unreadCount > 0 && props.initScrollOffsetFromBottom > 1000;
}
```

Comment (sridhar02) [reply components/toast_wrapper/toast_wrapper.jsx:87]: okay, I will change the logic.

Comment (sridhar02) [reply components/toast_wrapper/toast_wrapper.jsx]: Chnages Done

Comment (sudheerDev) [reply components/toast_wrapper/toast_wrapper.jsx]: Thank you, this looks good, Can you add a test case for less than 1000px?

Comment (sridhar02) [reply components/toast_wrapper/toast_wrapper.jsx]: yes I have added a new test case for less than 1000px

File: components/toast_wrapper/toast_wrapper.jsx [hunk 2]
```diff
@@ -85,7 +85,7 @@ class ToastWrapper extends React.PureComponent {
 
         // show unread toast on mount when channel is not at bottom and unread count greater than 0
         if (typeof showUnreadToast === 'undefined' && props.atBottom !== null) {
-            showUnreadToast = unreadCount > 0 && !props.atBottom;
```
Comment (Willyfrog) [inline components/toast_wrapper/toast_wrapper.jsx:88]: minor nitpick, can this `1000` scroll be a named constant so it can be more easily read?

**Comment by Willyfrog** [comment]
minor change, not even a blocking one

**Comment by esethna** [comment]
@sudheerDev @sridhar02 is this ready for re-review?

File: components/toast_wrapper/toast_wrapper.jsx (hunk 2 above)
Comment (sridhar02) [reply components/toast_wrapper/toast_wrapper.jsx:88]: yes done

**Comment by sridhar02** [comment]
@esethna yes, this PR is ready for the re-review.

**Comment by sridhar02** [comment]
@esethna if there is an issue with the PR, mention it, I will try to resolve it.

**Comment by esethna** [comment]
@jgilliam17 @sridhar02 as mentioned above I'm seeing the banner appear briefly before disappearing when the scroll to bottom is <1000px. I'd expect not to see the banner at all. cc// @sudheerDev 

https://i.gyazo.com/8727efd2508a0180e69f8fb94ace842a.gif

**Comment by jgilliam17** [comment]
@esethna yes, I can reproduce now. I missed that first time around. 
For me it reproduce only when manually marking posts as unread (using post menu option or alt+click), 
1. mark few posts as unread
2. leave the channel
3. return to the same channel again
environment: reproduces on both macOS and Windows
- Not able to reproduce the flashing toast if opening a channel when new messages are posted by other users.

**Comment by sridhar02** [comment]
@sudheerDev @esethna, I am not able to reproduce the issue in my local environment but I am able to see the issue in the URL given above using the mm-cloud-bot, I have also observed this in my local environment when i make messages as unread i am not able to see this message with orange color __new message__. I think this the line due to which we are getting the above issue of flashing of the toast under <1000px.
image

**Comment by sudheerDev** [comment]
@sridhar02 I can replicate this locally as well. I guess you were not able to replicate because you are marking same user messages as unread which does not show up the new messages line. 

Can you add couple of messages in the channel from a different user an then try?

File: components/post_view/post_list_virtualized/post_list_virtualized.test.jsx [hunk 3]
```diff
@@ -177,6 +177,11 @@ describe('PostList', () => {
                 scrollOffset: 490,
                 expected: true,
             },
+            {
+                name: 'when  pixel from the bottom also considered to be bottom',
```
Comment (sudheerDev) [inline components/post_view/post_list_virtualized/post_list_virtualized.test.jsx:181]: This test does not seem to add any value here as it is already. covered in the above assertion. Could you remove it?

**Comment by sudheerDev** [comment]
@sridhar02 is this ready for review again?

**Comment by sridhar02** [comment]
> @sridhar02 is this ready for review again?

I am stuck with the bug, I am not able to find out the cause, can you help @sudheerDev?

File: components/toast_wrapper/toast_wrapper.jsx [hunk 4]
```diff
@@ -90,7 +91,7 @@ class ToastWrapper extends React.PureComponent {
...
```
Comment (nevyangelova) [inline components/toast_wrapper/toast_wrapper.jsx:93]: Is there a reason to remove this check "!props.atBottom" ? I guess it might help resolving the bug?

Comment (sridhar02) [reply components/toast_wrapper/toast_wrapper.jsx:93]: previously it is present as 
```
 if (typeof showUnreadToast === 'undefined' && props.atBottom !== null) {
            showUnreadToast = unreadCount > 0 && !props.atBottom;
        }
````
i have changed to 
```
// show unread toast on mount when channel is not at bottom and unread count greater than 0
        if (typeof showUnreadToast === 'undefined' && props.atBottom !== null) {
            showUnreadToast = unreadCount > 0 && props.initScrollOffsetFromBottom > THRESHOLD_FROM_BOTTOM;
        }

```
which will get the solution for getting the toast from the bottom at 1000px

**Comment by esethna** [comment]
@sudheerDev can you help provide some guidance here, I'm seeing the same brief banner flash issue when marking a post as unread and then switching away and coming back to the channel. https://github.com/mattermost/mattermost-webapp/pull/5828#issuecomment-664535263

**Review by nevyangelova — APPROVED** [review]

**Comment by sudheerDev** [comment]
@sridhar02 

The mark as unread follows a different path in `getDerivedStateFromProps` You can check the couple of comments regarding the code flow for mark as unread

```
        if (props.atBottom !== null) {
            // show unread toast on mount when channel is not at bottom and unread count greater than 0
            if (typeof showUnreadToast === 'undefined') {
                showUnreadToast = unreadCount > 0 && props.initScrollOffsetFromBottom > THRESHOLD_FROM_BOTTOM;
            }

            if (typeof showMessageHistoryToast === 'undefined' && props.focusedPostId !== '') {
                showMessageHistoryToast = props.initScrollOffsetFromBottom > THRESHOLD_FROM_BOTTOM || !props.atLatestPost;
            }

            // show unread toast when a channel is marked as unread
            if (props.channelMarkedAsUnread && !prevState.channelMarkedAsUnread && !prevState.showUnreadToast) {
                showUnreadToast = props.initScrollOffsetFromBottom > THRESHOLD_FROM_BOTTOM;
            }

            // show unread toast when a channel is remarked as unread using the change in lastViewedAt
            // lastViewedAt changes only if a channel is remarked as unread in channelMarkedAsUnread state
            if (props.channelMarkedAsUnread && props.lastViewedAt !== prevState.lastViewedAt) {
                showUnreadToast = props.initScrollOffsetFromBottom > THRESHOLD_FROM_BOTTOM;
            }
        }
```

I changed the mark as unread paths to take into account `THRESHOLD_FROM_BOTTOM` (1000px) as well and as all. of these cases has `props.atBottom !== null` as common i changed that too.

Can you update the PR. with the changes and check if this works for you?

File: components/post_view/post_list_virtualized/post_list_virtualized.test.jsx (hunk 3 above)
Comment (sridhar02) [reply components/post_view/post_list_virtualized/post_list_virtualized.test.jsx]: resolved it

**Comment by sridhar02** [comment]
@sudheerDev I have checked with the logic, it is working fine now, can you review it now and I have removed unnecessary code.

**Review by sudheerDev — APPROVED** [review]
LGTM, thanks @sridhar02

**Comment by sudheerDev** [comment]
@sridhar02 Hey, Eric and I tested the spinmint and both of us still some weird issues, Let me check that again. Can you test the spinmint as well and see if everything is in order?

**Comment by stylianosrigas** [comment]
Removing the test server for maintenance purposes. Please add the label again if you need it. Sorry for any inconvenience.

**Comment by sridhar02** [comment]
@sudheerDev I have checked the application, I am not able to find any edge cases, toast for unread and marking as unread are working fine I think. are you able to find any edge cases for the toast, if found them what are they?

**Comment by sudheerDev** [comment]
@sridhar02 Changes seem. to work fine for me on spimint now 🤷‍♂️ . @jgilliam17 Can you test the PR again?

**Review by jgilliam17 — COMMENTED** [review]
Thank you @sridhar02 
The flashing toast previously reported is no longer reproducing.
I am seeing one new issue, where toast appears sometimes on the second try, not first
1. On the PR server login as sysadmin and visit Town Square (notice all the posts are from user1)
2. Scroll to the top and mark the first, top most, post as unread <- New message line, but no toast
3. Open a different channel and return to Town Square <- toast is present
4. Scroll to the bottom to clear the toast 
5. Repeat step 2.  <- toast is now present when marking the same post as unread

**Comment by sudheerDev** [comment]
> > I am seeing one new issue, where toast appears sometimes on the second try, not first
> > 
> > 1. On the PR server login as sysadmin and visit Town Square (notice all the posts are from user1)
> > 2. Scroll to the top and mark the first, top most, post as unread <- New message line, but no toast
> > 3. Open a different channel and return to Town Square <- toast is present
> > 4. Scroll to the bottom to clear the toast
> > 5. Repeat step 2.  <- toast is now present when marking the same post as unread
> 
> @sudheerDev Can you please confirm if this is the expected behavior? I know we discussed this previously, sorry I can't remember if we came to a consensus. Thank you.

That is not expected, Thoughts @esethna

**Comment by esethna** [comment]
@jgilliam17 I am not able to repro the above steps. I see the toast in step 2 & 5 as expected

**Comment by jgilliam17** [comment]
> @jgilliam17 I am not able to repro the above steps. I see the toast in step 2 & 5 as expected

Thank you Eric - I'll re-test.

**Comment by jgilliam17** [comment]
Thank you @sridhar02 
Tested, looks good to merge.

- No longer able to reproduce the [issue](https://github.com/mattermost/mattermost-webapp/pull/5828#pullrequestreview-488674018) mentioned previously (toast appearing only on 2nd attempt)
- Flashing toast previously reported is also no longer reproducing
- Verified new message toast shows when user's scroll position in center channel is 1000px or more from the bottom.

**Review by jgilliam17 — APPROVED** [review]

**Review by esethna — APPROVED** [review]

**Review by sudheerDev — APPROVED** [review]
LGTM

**Comment by sudheerDev** [comment]
@sridhar02 Hey Thanks for working on the feature, in future try to avoid force pushing the commits. Like in this case QA reviews are done and i see commit history changed and it would be hard to judge if you made any new changes after that
